File: java/src/main/java/org/sammancoaching/dependencies/Project.java
package org.sammancoaching.dependencies;

import static org.sammancoaching.dependencies.TestStatus.NO_TESTS;

public class Project {
    private final boolean deploysSuccessfully;
    private final TestStatus testStatus;
    private final boolean deploysSuccessfullyToStaging;
    private final TestStatus smokeTestStatus;

    public static ProjectBuilder builder() {
        return new ProjectBuilder();
    }

    private Project(boolean deploysSuccessfully, TestStatus unitTestStatus, boolean deploysSuccessfullyToStaging, TestStatus smokeTestStatus) {
        this.deploysSuccessfully = deploysSuccessfully;
        this.testStatus = unitTestStatus;
        this.deploysSuccessfullyToStaging = deploysSuccessfullyToStaging;
        this.smokeTestStatus = smokeTestStatus;
    }

    public String runTests() { return testStatus.getStatus(); }

    public String deploy() {
        return deploy(DeploymentEnvironment.PRODUCTION);
    }
    public String deploy(DeploymentEnvironment environment) {
        switch (environment) {
            case STAGING:
                return deploysSuccessfullyToStaging ? "success" : "failure";
            case PRODUCTION:
                return deploysSuccessfully ? "success" : "failure";
            default:
                return "failure";
        }
    }

    public TestStatus runSmokeTests() {
        return smokeTestStatus;
    }

    public static class ProjectBuilder {
        private boolean deploysSuccessfully;
        private TestStatus testStatus;
        private boolean deploysSuccessfullyToStaging = false;
        private TestStatus smokeTestStatus = NO_TESTS;

        public ProjectBuilder setTestStatus(TestStatus testStatus) {
            this.testStatus = testStatus;
            return this;
        }

        public ProjectBuilder setSmokeTestStatus(TestStatus smokeTestStatus) {
            this.smokeTestStatus = smokeTestStatus;
            return this;
        }

        public ProjectBuilder setDeploysSuccessfully(boolean deploysSuccessfully) {
            this.deploysSuccessfully = deploysSuccessfully;
            return this;
        }

        public ProjectBuilder setDeploysSuccessfullyToStaging(boolean deploysSuccessfully) {
            this.deploysSuccessfullyToStaging = deploysSuccessfully;
            return this;
        }

        public Project build() {
            return new Project(deploysSuccessfully, testStatus, deploysSuccessfullyToStaging, smokeTestStatus);
        }
    }
}

File: csharp/BuildPipeline/Project.cs
using System;

namespace UntangledConditionals
{
    public enum TestStatus {
        NoTests, //
        PassingTests, //
        FailingTests
    }
    
    public class Project
    {
        private readonly bool _deploysSuccessfully;
        private readonly TestStatus _testStatus;
        private readonly bool _deploysSuccessfullyToStaging;
        private readonly TestStatus _smokeTestStatus;

        public static ProjectBuilder Builder() {
            return new ProjectBuilder();
        }

        private Project(bool deploysSuccessfully, TestStatus testStatus, bool deploysSuccessfullyToStaging, TestStatus smokeTestStatus) {
            _deploysSuccessfully = deploysSuccessfully;
            _testStatus = testStatus;
            _deploysSuccessfullyToStaging = deploysSuccessfullyToStaging;
            _smokeTestStatus = smokeTestStatus;
        }

        public bool HasTests() {
            return _testStatus != TestStatus.NoTests;
        }

        public String RunTests() {
            return _testStatus == TestStatus.PassingTests ? "success" : "failure";
        }

        public String Deploy() {
            return Deploy(DeploymentEnvironment.Production);
        }

        public String Deploy(DeploymentEnvironment environment) {
            switch (environment) {
                case DeploymentEnvironment.Staging:
                    return _deploysSuccessfullyToStaging ? "success" : "failure";
                case DeploymentEnvironment.Production:
                    return _deploysSuccessfully ? "success" : "failure";
                default:
                    return "failure";
            }
        }

        public TestStatus RunSmokeTests() {
            return _smokeTestStatus;
        }

        public class ProjectBuilder {
            private bool _deploysSuccessfully;
            private TestStatus _testStatus;
            private bool _deploysSuccessfullyToStaging = false;
            private TestStatus _smokeTestStatus = TestStatus.NoTests;

            public ProjectBuilder SetTestStatus(TestStatus testStatus) {
                _testStatus = testStatus;
                return this;
            }

            public ProjectBuilder SetSmokeTestStatus(TestStatus smokeTestStatus) {
                _smokeTestStatus = smokeTestStatus;
                return this;
            }

            public ProjectBuilder SetDeploysSuccessfully(bool deploysSuccessfully) {
                _deploysSuccessfully = deploysSuccessfully;
                return this;
            }

            public ProjectBuilder SetDeploysSuccessfullyToStaging(bool deploysSuccessfully) {
                _deploysSuccessfullyToStaging = deploysSuccessfully;
                return this;
            }

            public Project Build() {
                return new Project(_deploysSuccessfully, _testStatus, _deploysSuccessfullyToStaging, _smokeTestStatus);
            }
        }
    }
}